doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,232 @@
+new monotonic append API on the `StorageController`.
+
+## Detailed description
+
+We'll update the parser to support creating a new kind of Source, `HTTP REQUESTS`, for which you can
+specify a `FORMAT`, which denotes how we decode the body of a request, the only supported format
+for the time being is `BYTES`.
+
+```
+CREATE SOURCE <name> FROM HTTP REQUESTS FORMAT BYTES;
+```
+
+After executing this request, we'll create a source with the following columns:
+
+| name    | type              |
+|---------|-------------------|
+| headers | map[text -> text] |
+| body    | bytea             |
+
+Using `bytea` might not be optimal from a storage perspective (e.g. no chance for MFP Pushdown) but
+it makes this initial attempt at webhooks maximally compatible. On the completion of the statement
+we'll return an ID to the user, which they would specify in the webhook URL. The returned ID would
+be 6 character base62 encoded string, a pattern commonly used in URL shorteners.
+
+> **_Note_**: An alternative to `bytea` encoded bodies would be to use `jsonb`. Looking through a
+few common apps that support webhooks, the payloads all default to JSON, or have JSON as the only
+option. Using `bytea` makes us compatible with all webhooks, with the tradeoff of adding work for
+the user to cast to their desired serialization format downstream.
+
+We'll add a new endpoint to the existing [`base_router`](https://github.com/MaterializeInc/materialize/blob/6e1f4c7352427301d782438d614feafb0f644442/src/environmentd/src/http.rs#L747)
+in `environmentd` with the path: `/api/webhook/:id`. This follows the existing pattern of our other
+two endpoints, whose paths also start with `/api`.
+
+> **_Note_**: The existing Web Sockets API is at `/api/experimental/ws`, we could move the webhooks
+API under the path `/api/experimental/webhook`, if we think there will be significant breaking
+changes in the future.
+
+We'll also add a new API `SessionClient::append_webhook` which is very similar to the existing
+[`insert_rows`](https://github.com/MaterializeInc/materialize/blob/6e1f4c7352427301d782438d614feafb0f644442/src/adapter/src/client.rs#L392-L397).
+This new API will send the captured `:id` from the URL path, the headers and body of the request,
+straight to the coordinator, via a new `Command`. In the coordinator we'll do the following:
+
+1. Spawn a new `tokio::task` to do the work off the main thread.
+2. Map the `id` to a previously created Source.
+3. Remove sensitive header values from the received headers (e.g. [Authorization](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization)).
+4. Decode the body of the request according to the Source's `FORMAT`.
+5. Pack the filtered headers and decoded body into a `Row`.
+6. Commit that `Row` with a new `StorageController::monotonic_append` API.
+7. Respond once the `StorageController` reports the data as committed.
```

Comment:
Step 1 moves off thread. Step 6 needs a storage controller, which must be back on thread. Step 7 wants to wait for it to complete. Assuming there aren't any correctness issues, can the await part here be done off thread also? That would mean the only times we're on thread are to get short lived access to the catalog and storage. We want to guarantee here that waiting for s3 writes doesn't block the main thread.

Reply:
Yeah I discovered during prototyping that step 6 needs to be on the coord thread. I updated the proposed `StorageController` API such that instead of the function taking a set of updates, it returns essentially a oneshot channel that we can use to send a batch of updates. This way we can do as little work as possible on the coord thread, and instead move nearly all of it into a separate `tokio::task`.

FWIW once I made this^ change, my prototype was able to support ~500 updates per-second, while there was a running subscribe in a separate shell, and querying system tables in a third.